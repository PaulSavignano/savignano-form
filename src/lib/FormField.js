import React, { useMemo } from 'react'
import PropTypes from 'prop-types'

import useFormField from './useFormField'

function FormField({
  name,
  id,
  onBlur,
  onChange,
  onFormat,
  onParse,
  onValidate,
  type,
  value,
  isPersistOnUnmount,
  component: Component,
  ...rest
}) {
  const fieldProps = useFormField({
    name,
    id,
    onBlur,
    onChange,
    onFormat,
    onParse,
    onValidate,
    isPersistOnUnmount,
    type,
    value,
  })
  return useMemo(() => {
    const { isTouched, ...restFieldProps } = fieldProps
    const isTouchedProp = typeof Component === 'string' ? {} : { isTouched }
    return (
      <Component {...rest} {...restFieldProps} {...isTouchedProp} />
    )
  }, [
    value,
    fieldProps.value,
    fieldProps.checked,
    fieldProps.error,
    fieldProps.isTouched,
  ])
}

FormField.defaultProps = {
  id: undefined,
  onBlur: undefined,
  onChange: undefined,
  onFormat: undefined,
  onParse: undefined,
  onValidate: undefined,
  type: undefined,
  value: undefined,
  isPersistOnUnmount: false
}

FormField.propTypes = {
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.node,
    PropTypes.string
  ]).isRequired,
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFormat: PropTypes.func,
  onParse: PropTypes.func,
  onValidate: PropTypes.oneOfType([PropTypes.func, PropTypes.array]),
  type: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.bool, PropTypes.number, PropTypes.string])
}

export default FormField
