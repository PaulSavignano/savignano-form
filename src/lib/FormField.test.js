import React from 'react'
import { mount, shallow } from 'enzyme'

import FormField from './FormField'
import FormContext from './FormContext'

const testCtx = {
  errors: {},
  onBlur: jest.fn(),
  onChange: jest.fn(),
  onRegisterField: jest.fn(),
  onUnregisterField: jest.fn(),
  touched: {},
  values: {}
}

const testProps = {
  component: 'input',
  id: '1',
  name: 'email',
  onBlur: undefined,
  onChange: undefined,
  onFormat: undefined,
  onParse: undefined,
  onValidate: undefined,
  type: 'text',
  value: undefined
}

describe('FormField', () => {
  it('should match snapshot', () => {
    const wrapper = mount(
      <FormContext.Provider value={testCtx}>
        <FormField {...testProps} />
      </FormContext.Provider>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
