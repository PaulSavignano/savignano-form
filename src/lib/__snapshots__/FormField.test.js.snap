// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField should match snapshot 1`] = `ReactWrapper {}`;
